flatgfa/src/packedseq.rs

```
@@ -0,0 +1,293 @@
+    } else {
+        let i: usize = (index / 2) as usize;
+        seq.data[i] = (0b11110000u8 & seq.data[i]) | elem;
+    }
+}
+
+/// Returns a uncompressed vector that contains the sequence in *seq*
+fn get_vec_seq(seq: &PackedVec) -> Vec<u8> {
+    let end_index = if (seq.high_nibble_end) {
+        (seq.data.len() * 2) - 1 
+    } else {
+        (seq.data.len() * 2) - 2
+    };
+    let span = Span {start: 0, end: end_index as u8};
+    return get_seq(&seq.data, span);
+}
+
+/// Returns a compressed PackedVec given an uncompressed vector *arr*
+fn create_vec(arr: Vec<u8>) -> PackedVec {
+    let mut high_nibble = true;
+    if arr.len() % 2 == 1 {
+        high_nibble = false;
+    }
+    let mut new_data = Vec::with_capacity(if high_nibble {
+        (arr.len() / 2) + 1
+    } else {
+        arr.len()
+    });
+    let mut j = 0;
+    let mut low = true;
+    for i in 0..arr.len() {
+        if low == true {
+            new_data.push(arr[i]);
+            low = false;
+        } else {
+            new_data[j] = new_data[j] | (arr[i] << 4);
+            low = true;
+            j += 1;
+        }
+    }
+    return PackedVec {
+        data: new_data,
+        high_nibble_end: high_nibble,
+    };
+}
+
+/// Returns a PackedSlice given a compressed PackVec *vec* that acts as a reference 
+/// to the section of *vec* contained within the index bounds of Span *s*.
+fn create_slice<'a>(vec: &'a PackedVec, s: Span) -> PackedSlice<'a> {
+    return PackedSlice{data_ref: &vec.data, span: s};
+}
+
+/// Returns a vector containing the base pairs referenced by *slice*
+fn get_slice_seq<'a>(slice: PackedSlice<'a>) -> Vec<u8> {
+    return get_seq(slice.data_ref, slice.span);
+}
+
+fn print_arr(arr: Vec<u8>) {
+    let new_arr: Vec<char> = arr
+        .into_iter()
+        .map(|e| match e {
+            0 => 'A',
+            1 => 'C',
+            2 => 'T',
+            3 => 'G',
+            _ => ' ',
```

I say let's make this an error: as in, use `panic!` to crash if we ever get an unsupported value.